Reduce size of images in theme picker

danluu opened the pull request:
I'm new to Calypso and Wordpress and am basically a contractor or intern who doesn't know anything, so apologies if I'm missing a step somewhere in the process here. This is a follow-up to @niranjan-uma-shankar finding that the design picker has a 70 MB payload. Looking at the funnel data, we can see that a lot of users drop out at this step, including new users. When segmenting users, we can see that users are relatively more likely to drop out if they have a slow connection or a slow device, which makes sense considering the payload size. From looking at the history here, it looks like there was a change made in 2017 by @fditrapani that reduced the number of themes shown in a category from 9 to 3 (compared to 84 now in the default category), which reduced the number of users dropping out of this flow, especially on mobile. That seems related, in that page weight and complexity tend to cause more attrition on mobile since users are more likely to be on devices or connections that can't handle a heavy or complex plage.

There are a number of ways we could try to address this, including A/B testing various changes, but @bazza suggested that it would be ok to just try deploying a simple change that reduces image sizes and then looking at our funnels to see the impact. This commit reduces the payload size to 16 MB, which is still too large a payload for many users, but will allow more users to be able to pick a theme and create a site.

Without going back to static images, another solution would be to modify mShots to allow webp output or to use photon to get a smaller webp, but just doing that is probably not sufficient to get image sizes down to a level where almost every user will be able to use the theme selector, just like this change alone isn't large enough to do that. Perhaps both of those combined or one of those combined with having fewer themes shown by default could work.

## Proposed Changes

This change reduces the size of each thumbail that's downloaded from `1199x3600` to `500x1000`. The `1000` seems wasteful in that, with most window sizes a user might have, the thumbnail will be displayed at a much lower resolution (for example, fullscreened, on a `3440x1440` display, the actual size of each thumbnail is `460x300`. If I drag my window to try a bunch of possible sizes, I can get a bigger thumbnail, up to something like `600x400`, but `400` is still much smaller than `1000`. But, due to the way the images are handled, without tweaking other settings, much less of the page is shown if a height of less `1000` is picked and I don't think I understand the setup here enough to pick a much better size. With jpeg images, the total size of the image at a fixed quality for an image like this seems like it should be super-linear in the number of pixels, so most data the user downloads is effectively being thrown away even with the smaller version here.

Someone on a retina display can probably observe slightly lower image quality, but if the tradeoff for better image quality on a retina display is that many users are unable to load this page and pick a theme, I'm not sure that's worth it.

Below are screenshots of the original and the changed version on a `3440x1440` display:

![designsetup-a](https://github.com/Automattic/wp-calypso/assets/157136/546be572-7fa8-4470-9232-f68a8860167d)

![designsetup-b](https://github.com/Automattic/wp-calypso/assets/157136/8f4d142d-0e41-48e5-95e4-6fdf5acdcc02)

If we zoom in on both versions, looking at a few different parts, we can see that almost all of the loss in image quality comes from scaling artifacts and that throwing more pixels at the user has little to no effect on image quality (in some areas, the `1199x3600` image looks slightly better, and in some cases, the `500x1000` image looks slightly better. Which parts look better has more to do with how the small displayed size causes aliasing with respect to the "raw" sisze than the number of pixels in the raw large image.

![zoom-1-a](https://github.com/Automattic/wp-calypso/assets/157136/04dc40f4-c682-4d6e-a68b-9bbb09dc56fd)

![zoom-1-b](https://github.com/Automattic/wp-calypso/assets/157136/764fb15a-93b4-4678-b4b3-3e33327783ce)

![zoom-2-a](https://github.com/Automattic/wp-calypso/assets/157136/02eff1b7-a30f-4b32-ac0d-39f8f9cf16eb)

![zoom-2-b](https://github.com/Automattic/wp-calypso/assets/157136/8a8fc18b-0d05-4517-a95e-b370516423bf)


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* This change is observable by setting up a new site and navigating to `http://calypso.localhost:3000/setup/update-design/designSetup?siteSlug={yoursitehere}` during the process.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

danluu committed:
Reduce size of images in designSetup by passing smaller arguments to
mShot.

This is a follow-up to @niranjan-uma-shankar finding that the design
picker has a 70 MB payload. We can see in our funnel data that this
causes a lot of users, including new users, to drop the process of
creating a site at this step. When segmenting users, we can see that
users are relatively more likely to drop out if they have a slow
connection or a slow device, which supports the idea that 70 MB is too
large a payload for many users.

This commit reduces the payload size to 16 MB, which is still too large
a payload for many users, but will allow more users to be able to pick
a theme and create a site. There are a number of possible longer-term
solutions that could get this payload down to the point where almost
every user will be able to pick a theme, but those will be more involved
fixes.

dsas commented:
I think the quality difference is noticeable. I'm using my 2056 * 1329 screen resolution and I can go from sharp, legible text to slightly mushy sometimes illegible text.

https://github.com/Automattic/wp-calypso/assets/93301/91db0158-e0d0-4037-8e37-98d3c805e702

danluu commented:
> I think the quality difference is noticeable. I'm using about half of my monitor with a 2056 * 1329 screen resolution and I can go from sharp, legible text to slightly mushy sometimes illegible text.

What do you need to do to get a two-column wide format? If I make my window < `780px` or so, I get a single column, and then above that, I get 3 columns. With 3 columns, in some cases, the original looks a bit better and in others, the new version looks better. With 1 column, the original looks better until the window is pushed down to be smaller, maybe `600px` or `500px` or so.

I'm using Chrome on Mac, current version of both, with ad blocker disabled and nothing else that should be changing the layout of a page.

dsas commented:
@danluu oh sorry for the confusion that isn't a two column wide format, it's just a partial screen capture

I'm using firefox on mac fwiw.

dsas commented on packages/design-picker/src/utils/available-designs.ts:
```diff
@@ -4,25 +4,24 @@ import type { MShotsOptions } from '@automattic/onboarding';
-		w: highRes ? 1199 : 600,
-		screen_height: 3600,
+		w: 500,
+		screen_height: 1000,
```
I think changing the screen_height is displaying more of the page than before. This probably needs to be at least 1100 to use the same size.

danluu replied:
Yeah, I wasn't sure what to do about this. On a 1x display, any image over `500` to maybe `600` tall is actually wasted page weight, but the scaling we do means that the image is distorted even at `1000`, which is why, when you toggle between the two options in the video, one shows more of the page than the other (and this is also visible in my comparison, where one of the two versions shows more of the page).

I don't understand what's going on here enough to know why we need this height that's close to double the actual displayed resolution (on a 1x display) for the screenshot to display correctly.

dsas replied:
I don't fully understand it either tbh (perhaps @Automattic/lego might have some ideas). We definitely don't need the full 3600 though.

danluu replied:
I think the 3600 is for a "long previews" feature. I saw this when I was looking at the history here to see how we got here (apparently previews used to be `600x600`. It seems like, when you mouseover, the thumbnail is supposed to scroll and I saw some videos of this happening, but I wasn't able to trigger this scrolling behavior myself. I'm not sure what's necessary to enable the behavior, but my browser definitely doesn't need a `3600` tall image if most of it is permanently off screen, and even in the cases where the behavior triggers correctly, I'd be interested in seeing the data about the lift we get from having the feature vs. the drag we get from some users not being able to use the page.

[sorry, I originally left this comment in the wrong spot since I replied by email, and I guess that automatically appends your comment to the main issue even when you're replying to a comment on some code]

danluu replied:
On looking at the actual images that get served up, requesting `w=1199` and `height=3600` actually gives us an image that's `1199x2697`, at least for one of the images. For the arguments in the new version `w=500`, `height=1000`, we get an image that's `500x375`. So passing a huge `height` doesn't necessarily give us a bunch of pixels which users mostly won't see.

danluu commented:
@dsas That makes sense, thanks! `2056 x 1329` is a 16" Macbook with a retina display? If that's with the default 2x native scaling, I think it makes sense that there's a more observable difference on a retina display because you're now actually losing resolution. On my higher resolution but non-retina 1x display, the scaling artifacts are coming from scaling an `1199px` image or a `500px` image into (unless the window is carefully selected to avoid this) something like `450px`.

If the above is wrong, then that's good information for me. If it's not wrong then, IMO, it's not worth making things look nicer for people on retina displays if this means a large fraction of users are served a page that can't really interact with. But, for preserving crispness for people on retina displays, maybe we could detect `devicePixelRatio` and try to serve up higher resolution images if people are on 2x displays? I'm not sure that's the right idea, since the page will still be unusable if someone has a semi-modern macbook and a poor internet connection, but it would look nicer in the cases where the user's connection is fast enough to download the higher res images.

dsas commented:
Yes that's my 16" retina macbook, but I'm using my external monitor with a 3008*1692 resolution (scaled down from native 4k display), doesn't change much though.

Yeah, I don't want to block this. I think we should give it a try and see what difference it makes to the numbers. We can then decide whether it's worth keeping, or investing more time in a lazy loading fix or something. 

The commit checklist (PCYsg-hS-p2) does say "If there are images involved, they must be HiDPI optimized. (Retinaaaaaaaaa!)" but rules are there for the breaking 🤫

danluu commented:
I think the 3600 might be for some kind of “long previews” functionality? I
saw some videos that I think demonstrate this, where the image scrolls on
mouseover, but I wasn’t able to trigger that myself and I’m not sure why
that is.

On Fri, Mar 22, 2024 at 1:12 PM Dean Sas ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In packages/design-picker/src/utils/available-designs.ts
> <https://github.com/Automattic/wp-calypso/pull/88786#discussion_r1536150942>
> :
>
> >  		vph: scrollable ? 1600 : 1040,
> -		// When `w` was 1200 it created a visual glitch on one thumbnail. #57261
> -		w: highRes ? 1199 : 600,
> -		screen_height: 3600,
> +		w: 500,
> +		screen_height: 1000,
>
> I don't fully understand it either tbh (perhaps @Automattic/lego
> <https://github.com/orgs/Automattic/teams/lego> might have some ideas).
> We definitely don't need the full 3600 though.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Automattic/wp-calypso/pull/88786#discussion_r1536150942>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABGLUCUOI5LSQTAYABK7VLYZSGCNAVCNFSM6AAAAABFCWRNWSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSNJVG4YTCNBRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

danluu commented:
That makes sense. On lazy loading in particular, my experience has been that lazy loading usually makes my experience worse if I have a slow connection and it often prevents me from viewing the page at all, even if I'm patient, if I'm using a slow device. By slow device here, I mean something like a Tecno Spark 8C, which, income adjusted, costs more in many countries where this device is sold (India, Nigeria, etc.) than a new iPhone costs in the U.S.

As another experiment, here's one of the images run through [Site Accelerator / Photon](https://developer.wordpress.com/docs/site-accelerator/):

[Version above (new, not original)](https://s0.wp.com/mshots/v1/https%3A%2F%2Fpublic-api.wordpress.com%2Fwpcom%2Fv2%2Fblock-previews%2Fsite%3Fstylesheet%3Dpub%252Fissue%26language%3Den%26viewport_height%3D1040%26use_screenshot_overrides%3Dtrue%26site_title%3DIssue?vpw=1600&vph=1040&w=500&screen_height=1000&count=0), `46K`
[After photon](https://i0.wp.com/s0.wp.com/mshots/v1/https%3A%2F%2Fpublic-api.wordpress.com%2Fwpcom%2Fv2%2Fblock-previews%2Fsite%3Fstylesheet%3Dpub%252Fissue%26language%3Den%26viewport_height%3D1040%26use_screenshot_overrides%3Dtrue%26site_title%3DIssue?vpw=1600&vph=1040&w=500&screen_height=1000&count=0), `34K`.

I was hoping Photon would give me a `webp` since my browser supports it, but I get a `jpeg` back. 

```
%  curl -s -H "Accept: image/webp" -I "https://i0.wp.com/s0.wp.com/mshots/v1/https%3A%2F%2Fpublic-api.wordpress.com%2Fwpcom%2Fv2%2Fblock-previews%2Fsite%3Fstylesheet%3Dpub%252Fissue%26language%3Den%26viewport_height%3D1040%26use_screenshot_overrides%3Dtrue%26site_title%3DIssue?vpw=1600&vph=1040&w=500&screen_height=1000&count=0&quality=70" | grep ^content-type 

content-type: image/jpeg
```

We can get a smaller image by passing a quality argument. With [quality=70](https://i0.wp.com/s0.wp.com/mshots/v1/https%3A%2F%2Fpublic-api.wordpress.com%2Fwpcom%2Fv2%2Fblock-previews%2Fsite%3Fstylesheet%3Dpub%252Fissue%26language%3Den%26viewport_height%3D1040%26use_screenshot_overrides%3Dtrue%26site_title%3DIssue?vpw=1600&vph=1040&w=500&screen_height=1000&count=0&quality=70), we get `20K`. Cutting these image sizes in half still leaves us with a payload that's too large for many users, but it's an improvement. There's something non-ideal about this, since `mShots` looks like it's giving us an image created with `gd-jpeg` at `95` quality, which we're then trying to compress again at `70` quality. We'd get a smaller and higher quality image if we could get lossless version of the original image or a version at `100` quality that we then compress to `70` quality but, in terms of the overall experience for the user, I'd still expect this to be an improvement.

danluu committed:
Increase image size slightly over previous experiment

This requests 600x1100, which should give us 100% visual fidelity on
a 1x display. This will have some perceptual difference in quality
on a 2x/retina display.

dsas reviewed (approved):
This makes sense to me, there's a tiny drop in image quality for a substantial drop in file size. I'm interested to see what happens to the onboarding funnels.

danluu commented:
Yeah, I'm curious what the numbers look like. To see if we can also easily photon-ize the images in the same PR, I tried adding photon since it seems like we can get a 50% to 75% reduction in image size with only a small reduction in image quality (most perceptible when we have sharp lines, as in the themes that have big block text, not really perceptible without zooming in when there are images). This had three problems, listed here, most problematic first:

* For some reason, some images get stuck on "generating preview" indefinitely. Refreshing many times after long delays doesn't fix this. Perhaps photon is caching the loading animation?
* Photon and mshots "want to" consume the same arguments. Since photon has to be applied first (we have a url that looks like `i0.wp.com/s0.wp.com/...`, photon consumes all of the arguments, so arguments that are intended to be passed to mshot don't get passed and we get default arguments to mshot, which changes what's displayed in the images. In many cases, I don't think this makes any difference, but this gives us less vertical space, so for blogs which have (for example) a big header and then some text, we may not see what the text looks like.
* As mentioned above, we get `jpeg` and not `webp` images even when we have `Accept: image/webp`.

On the first issue, I think this is basically fatal without a fix. Here are the headers from one image where it seems like (?) photon has cached the loading animaton:

```
%  curl -s -H "Accept: image/webp" -I "https://i0.wp.com/s0.wp.com/mshots/v1/https%3A%2F%2Fpublic-api.wordpress.com%2Fwpcom%2Fv2%2Fblock-previews%2Fsite%3Fstylesheet%3Dpub%252Fskatepark%26language%3Den%26viewport_height%3D1040%26use_screenshot_overrides%3Dtrue%26site_title%3DSkatepark?vpw=1600&vph=1040&w=600&screen_height=1100&quality=70&count=0"
HTTP/2 200 
server: nginx
date: Tue, 26 Mar 2024 00:51:40 GMT
content-type: image/gif
content-length: 8737
last-modified: Tue, 26 Mar 2024 00:39:57 GMT
expires: Thu, 26 Mar 2026 12:39:57 GMT
cache-control: public, max-age=63115200
link: <http://s0.wp.com/mshots/v1/https%3A%2F%2Fpublic-api.wordpress.com%2Fwpcom%2Fv2%2Fblock-previews%2Fsite%3Fstylesheet%3Dpub%252Fskatepark%26language%3Den%26viewport_height%3D1040%26use_screenshot_overrides%3Dtrue%26site_title%3DSkatepark>; rel="canonical"
x-content-type-options: nosniff
etag: "e7bd96bc9167451c"
vary: Accept
x-nc: HIT sea 8
alt-svc: h3=":443"; ma=86400
access-control-allow-origin: *
access-control-allow-methods: GET, HEAD
timing-allow-origin: *
```

Unless someone knows an easy fix or workaround for this, I think it makes sense to defer trying to wrap mshots in photon to another PR.

danluu committed:
Switch back to 500 wide, but keep 1100 depth

dsas commented:
> From running the tests to see why tests failed earlier, it seems like tests are quite flaky and may randomly pass or fail?

Generally speaking, the unit tests are pretty solid and the e2e tests are often flaky (my first instinct is to just re-run them rather than investigate).

BTW, since you've made this change via a fork I keep manually re-running the GitHub builds (PCYsg-t1d-p2), if you just push a branch to this repo it will happen automatically.

danluu commented:
Thanks for the info!

One thing I'll mention for posterity is that, for the images that are currently using photon ("partner" themes, which use static screenshots instead of mShots), adding `quality=70` generally didn't help. It sometimes made no difference and sometimes made the image larger, as in this case:

Original:

```
%  curl -s -I "https://i0.wp.com/theme.files.wordpress.com/2023/01/makoney-preview-2.jpg?ssl=1&fit=479%2C360&zoom=2" | rg content-length 

content-length: 50010
```

With `70` quality:

```
curl -s -I "https://i0.wp.com/theme.files.wordpress.com/2023/01/makoney-preview-2.jpg?ssl=1&fit=479%2C360&zoom=2&quality=70" | rg content-length

content-length: 127932
```

I didn't check every image, so there may be images where this makes the image smaller as well, but this doesn't seem worth adding to the static screenshots.

It's outside of the scope of this issue, but I don't think that passing a low quality argument should cause Photon to produce a larger image.

danluu commented:
Whoops, on looking at what the browser actually loads and not just what we get with `curl`, the above comment is correct but not relevant for most users.

If we accept `webp`, which almost all browsers do in the wild, we get a significantly larger image. This is another thing that shouldn't happen, but in terms of a quick and easy fix, we can serve up static images that are consistent with `mShot` images in terms of display size (I think that makes sense to do because, on the default load page, 2 out of 82 themes use static images and those 2 themes now account for something like 12% of all image weight). I'll make a separate PR for that because this is now merged.
